[JIT] greedy by size memory planning strategy

makslevental opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63873
* #64350
* __->__ #64349
* #64348
* #64347
* #62995

Greedy by size (a heuristic that allocates largest tensors first) based on [this paper]( https://arxiv.org/pdf/2001.03288.pdf). It's factored the way it is for the purposes of making subsequent strategies simpler to implement.

make greedy deterministic

Differential Revision: [D30769096](https://our.internmc.facebook.com/intern/diff/D30769096)

Commit:
[JIT] greedy by size memory planning strategy

Greedy by size (a heuristic that allocates largest tensors first) based on [this paper]( https://arxiv.org/pdf/2001.03288.pdf). It's factored the way it is for the purposes of making subsequent strategies simpler to implement.

make greedy deterministic

[ghstack-poisoned]

Commit:
Update on "[JIT] greedy by size memory planning strategy"


Greedy by size (a heuristic that allocates largest tensors first) based on [this paper]( https://arxiv.org/pdf/2001.03288.pdf). It's factored the way it is for the purposes of making subsequent strategies simpler to implement.

make greedy deterministic

[ghstack-poisoned]

makslevental committed:
Update on "[JIT] greedy by size memory planning strategy"


Greedy by size (a heuristic that allocates largest tensors first) based on [this paper]( https://arxiv.org/pdf/2001.03288.pdf). It's factored the way it is for the purposes of making subsequent strategies simpler to implement.

make greedy deterministic

[ghstack-poisoned]

Commit:
Update on "[JIT] greedy by size memory planning strategy"


Greedy by size (a heuristic that allocates largest tensors first) based on [this paper]( https://arxiv.org/pdf/2001.03288.pdf). It's factored the way it is for the purposes of making subsequent strategies simpler to implement.

make greedy deterministic

[ghstack-poisoned]

makslevental committed:
Update on "[JIT] greedy by size memory planning strategy"


Greedy by size (a heuristic that allocates largest tensors first) based on [this paper]( https://arxiv.org/pdf/2001.03288.pdf). It's factored the way it is for the purposes of making subsequent strategies simpler to implement.

make greedy deterministic

[ghstack-poisoned]

Commit:
Update on "[JIT] greedy by size memory planning strategy"


Greedy by size (a heuristic that allocates largest tensors first) based on [this paper]( https://arxiv.org/pdf/2001.03288.pdf). It's factored the way it is for the purposes of making subsequent strategies simpler to implement.

make greedy deterministic

[ghstack-poisoned]

Commit:
Update on "[JIT] greedy by size memory planning strategy"


Greedy by size (a heuristic that allocates largest tensors first) based on [this paper]( https://arxiv.org/pdf/2001.03288.pdf). It's factored the way it is for the purposes of making subsequent strategies simpler to implement.

make greedy deterministic

[ghstack-poisoned]

Commit:
Update on "[JIT] greedy by size memory planning strategy"


Greedy by size (a heuristic that allocates largest tensors first) based on [this paper]( https://arxiv.org/pdf/2001.03288.pdf). It's factored the way it is for the purposes of making subsequent strategies simpler to implement.

make greedy deterministic

[ghstack-poisoned]

Commit:
Update on "[JIT] greedy by size memory planning strategy"


Greedy by size (a heuristic that allocates largest tensors first) based on [this paper]( https://arxiv.org/pdf/2001.03288.pdf). It's factored the way it is for the purposes of making subsequent strategies simpler to implement.

make greedy deterministic

[ghstack-poisoned]

makslevental commented:
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30769096).

Commit:
Update on "[JIT] greedy by size memory planning strategy"


Greedy by size (a heuristic that allocates largest tensors first) based on [this paper]( https://arxiv.org/pdf/2001.03288.pdf). It's factored the way it is for the purposes of making subsequent strategies simpler to implement.

make greedy deterministic

Differential Revision: [D30769096](https://our.internmc.facebook.com/intern/diff/D30769096)

[ghstack-poisoned]

Commit:
Update on "[JIT] greedy by size memory planning strategy"


Greedy by size (a heuristic that allocates largest tensors first) based on [this paper]( https://arxiv.org/pdf/2001.03288.pdf). It's factored the way it is for the purposes of making subsequent strategies simpler to implement.

make greedy deterministic

Differential Revision: [D30769096](https://our.internmc.facebook.com/intern/diff/D30769096)

[ghstack-poisoned]

Commit:
Update on "[JIT] greedy by size memory planning strategy"


Greedy by size (a heuristic that allocates largest tensors first) based on [this paper]( https://arxiv.org/pdf/2001.03288.pdf). It's factored the way it is for the purposes of making subsequent strategies simpler to implement.

make greedy deterministic

Differential Revision: [D30769096](https://our.internmc.facebook.com/intern/diff/D30769096)

[ghstack-poisoned]

Commit:
Update on "[JIT] greedy by size memory planning strategy"


Greedy by size (a heuristic that allocates largest tensors first) based on [this paper]( https://arxiv.org/pdf/2001.03288.pdf). It's factored the way it is for the purposes of making subsequent strategies simpler to implement.

make greedy deterministic

Differential Revision: [D30769096](https://our.internmc.facebook.com/intern/diff/D30769096)

[ghstack-poisoned]

Commit:
Update on "[JIT] greedy by size memory planning strategy"


Greedy by size (a heuristic that allocates largest tensors first) based on [this paper]( https://arxiv.org/pdf/2001.03288.pdf). It's factored the way it is for the purposes of making subsequent strategies simpler to implement.

make greedy deterministic

Differential Revision: [D30769096](https://our.internmc.facebook.com/intern/diff/D30769096)

[ghstack-poisoned]

Commit:
Update on "[JIT] greedy by size memory planning strategy"


Greedy by size (a heuristic that allocates largest tensors first) based on [this paper]( https://arxiv.org/pdf/2001.03288.pdf). It's factored the way it is for the purposes of making subsequent strategies simpler to implement.

make greedy deterministic

Differential Revision: [D30769096](https://our.internmc.facebook.com/intern/diff/D30769096)

[ghstack-poisoned]

makslevental committed:
Update on "[JIT] greedy by size memory planning strategy"


Greedy by size (a heuristic that allocates largest tensors first) based on [this paper]( https://arxiv.org/pdf/2001.03288.pdf). It's factored the way it is for the purposes of making subsequent strategies simpler to implement.

make greedy deterministic

Differential Revision: [D30769096](https://our.internmc.facebook.com/intern/diff/D30769096)

[ghstack-poisoned]

makslevental committed:
Update on "[JIT] greedy by size memory planning strategy"


Greedy by size (a heuristic that allocates largest tensors first) based on [this paper]( https://arxiv.org/pdf/2001.03288.pdf). It's factored the way it is for the purposes of making subsequent strategies simpler to implement.

make greedy deterministic

Differential Revision: [D30769096](https://our.internmc.facebook.com/intern/diff/D30769096)

[ghstack-poisoned]

makslevental committed:
Update on "[JIT] greedy by size memory planning strategy"


Greedy by size (a heuristic that allocates largest tensors first) based on [this paper]( https://arxiv.org/pdf/2001.03288.pdf). It's factored the way it is for the purposes of making subsequent strategies simpler to implement.

make greedy deterministic

Differential Revision: [D30769096](https://our.internmc.facebook.com/intern/diff/D30769096)

[ghstack-poisoned]

makslevental committed:
Update on "[JIT] greedy by size memory planning strategy"


Greedy by size (a heuristic that allocates largest tensors first) based on [this paper]( https://arxiv.org/pdf/2001.03288.pdf). It's factored the way it is for the purposes of making subsequent strategies simpler to implement.

make greedy deterministic

Differential Revision: [D30769096](https://our.internmc.facebook.com/intern/diff/D30769096)

[ghstack-poisoned]

makslevental committed:
Update on "[JIT] greedy by size memory planning strategy"


Greedy by size (a heuristic that allocates largest tensors first) based on [this paper]( https://arxiv.org/pdf/2001.03288.pdf). It's factored the way it is for the purposes of making subsequent strategies simpler to implement.

make greedy deterministic

Differential Revision: [D30769096](https://our.internmc.facebook.com/intern/diff/D30769096)

[ghstack-poisoned]

Krovatkin commented on torch/csrc/jit/passes/memory_planning/greedy_util.cpp:
when would this case happen you seem to be inserting in the order of increasing offsets in `makeAllocation`?

makslevental replied:
@Krovatkin good catch - it's true this doesn't make sense in this function (`findFirstOffset`) but it has since been refactored and makes more sense in context

Krovatkin commented on torch/csrc/jit/passes/memory_planning/greedy_util.cpp:
```diff
@@ -0,0 +1,83 @@
+  size_t prev_offset = 0;
+
+  for (const auto& alloc : ordered_allocations) {
+    if (!overlapLiveRange(alloc.ulvr, unalloced_ulvr)) {
```
I'm a bit confused here. It looks like we are skipping live ranges that don't overlap with the unallocated live range? 
I was thinking we should be specifically looking for  the ranges that do *not* overlap otherwise we can't reuse that storage?

makslevental replied:
@Krovatkin this is looking for live ranges that overlap in time and then finding the smallest gap *between* these "live" live ranges. this should really be done using something like an interval tree (and an overlap lookup). I'm looking into this but one issue is interval trees don't deal well with duplicated endpoints. In the single-threaded case that's not an issue but in the concurrency case it might pop up.

Krovatkin reviewed (commented):
a few qs. I seem to be struggling to fully understand this PR.

makslevental commented on torch/csrc/jit/passes/memory_planning/greedy_by_size.cpp:
```diff
@@ -0,0 +1,126 @@
+    makeAllocation(ulvr, size, ordered_allocations, findOffsetWithSmallestGap);
+  }
+
+  auto cmp = liveRangeStartCmp();
```
@Krovatkin that's correct

makslevental commented on torch/csrc/jit/passes/memory_planning/greedy_by_size.cpp:
```diff
@@ -0,0 +1,126 @@
+      [&cmp](auto m1, auto m2) { return cmp(m1.ulvr, m2.ulvr); });
+  return ordered_allocations;
+}
+
```
@Krovatkin hmm not sure what version you were looking at but it's in this commit

![image](https://user-images.githubusercontent.com/5657668/140011131-058be125-93cb-4b4e-8f6f-7cef33a7fd43.png)

makslevental commented on torch/csrc/jit/passes/memory_planning/greedy_by_size.cpp:
```diff
@@ -0,0 +1,126 @@
+      [&cmp](auto& p1, auto& p2) {
+        auto len1 = p1.first.lvr.end - p1.first.lvr.begin;
+        auto len2 = p2.first.lvr.end - p2.first.lvr.begin;
+        return len1 == len2 ? (p1.second == p2.second ? cmp(p1.first, p2.first)
```
`*.first` are the "unique" live ranges. we're sorting here by length and then breaking ties by size and start. it's convoluted i admit but it happens and it's very frustrating dealing with non-deterministic orderings when debugging.

makslevental committed:
Update on "[JIT] greedy by size memory planning strategy"


Greedy by size (a heuristic that allocates largest tensors first) based on [this paper]( https://arxiv.org/pdf/2001.03288.pdf). It's factored the way it is for the purposes of making subsequent strategies simpler to implement.

make greedy deterministic

Differential Revision: [D30769096](https://our.internmc.facebook.com/intern/diff/D30769096)

[ghstack-poisoned]

makslevental commented:
@makslevental has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30769096).
